source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+            left_sphere_r_item.topology = renderer::PrimitiveTopology::TRIANGLELIST;
+            left_sphere_r_item.index_count = left_sphere_r_item.geometry->draw_args["sphere"_small].index_count;
+            left_sphere_r_item.start_index_location = left_sphere_r_item.geometry->draw_args["sphere"_small].start_index_location;
+            left_sphere_r_item.base_vertex_location = left_sphere_r_item.geometry->draw_args["sphere"_small].base_vertex_location;
+            left_sphere_r_item.num_frames_dirty = renderer::num_frames_in_flight();
+
+            right_sphere_r_item.world = right_sphere_world;
+            right_sphere_r_item.constant_buffer_index = obj_cb_index++;
+            right_sphere_r_item.geometry = g_regina_ctx.meshes["scene_geometry"_med].get();
+            right_sphere_r_item.topology = renderer::PrimitiveTopology::TRIANGLELIST;
+            right_sphere_r_item.index_count = right_sphere_r_item.geometry->draw_args["sphere"_small].index_count;
+            right_sphere_r_item.start_index_location = right_sphere_r_item.geometry->draw_args["sphere"_small].start_index_location;
+            right_sphere_r_item.base_vertex_location = right_sphere_r_item.geometry->draw_args["sphere"_small].base_vertex_location;
+            right_sphere_r_item.num_frames_dirty = renderer::num_frames_in_flight();
+
+            g_regina_ctx.scene->add_render_item(rsl::move(left_cyl_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(right_cyl_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(left_sphere_r_item));
+            g_regina_ctx.scene->add_render_item(rsl::move(right_sphere_r_item));
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_frame_resources()
+    {
+        for (int i = 0; i < renderer::num_frames_in_flight(); ++i)
+        {
+            renderer::ResourceSlot frame = renderer::create_frame_resource();
+
+            g_regina_ctx.frame_resource_data.push_back({rsl::move(frame) , renderer::ResourceSlot::make_invalid(), renderer::ResourceSlot::make_invalid()});
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_constant_buffers()
+    {
+        s32 num_render_items = g_regina_ctx.scene->render_item_count();
+
+        for (s32 frame = 0; frame < renderer::num_frames_in_flight(); ++frame)
+        {
+            renderer::commands::AttachCommitedResourceToFrameCommandDesc attach_object_constants;
+            attach_object_constants.frame_slot = renderer::frame_at_index(frame);
+            attach_object_constants.buffer_count = num_render_items;
+            attach_object_constants.buffer_byte_size = sizeof(ObjectConstants);
+
+            g_regina_ctx.frame_resource_data[frame].object_commited_resource = renderer::attach_commited_resource_to_frame(rsl::move(attach_object_constants));
```

Comment:
assigning to member of frame resource data. I guess it's fine for now, but ideally we don't want to do this, this should be a member function of t he frame resource data instead. You know how many frames in flight you can have, you know the number of render items you have and the size of a per object and per pass cb (although this should be managed by the engine, not the editor imo). so you can pass in that data into a member function and let the frame resource data handle it internally